compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalJvmCompilerRunner.kt

```diff
@@ -456,24 +470,118 @@ class IncrementalJvmCompilerRunner(
 
     override fun performWorkAfterSuccessfulCompilation(caches: IncrementalJvmCachesManager) {
         if (classpathChanges is ClasspathChanges.ClasspathSnapshotEnabled) {
-            reporter.measure(BuildTime.SAVE_SHRUNK_CURRENT_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
+            reporter.measure(BuildTime.SHRINK_AND_SAVE_CURRENT_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
                 shrinkAndSaveClasspathSnapshot(classpathChanges, caches.lookupCache)
             }
         }
     }
 
+    /** Used by [shrinkAndSaveClasspathSnapshot]. */
+    private sealed class ShrinkMode {
+        object NoChanges : ShrinkMode()
+
+        class IncrementalNoNewLookups(
+            val shrunkCurrentClasspathAgainstCurrentLookups: List<ClassSnapshotWithHash>,
+        ) : ShrinkMode()
+
+        class Incremental(
+            val currentClasspathSnapshot: List<ClassSnapshotWithHash>,
+            val shrunkCurrentClasspathAgainstPreviousLookups: List<ClassSnapshotWithHash>,
+            val addedOrUpdatedLookupSymbols: Set<LookupSymbolKey>
+        ) : ShrinkMode()
+
+        object NonIncremental : ShrinkMode()
+    }
+
     private fun shrinkAndSaveClasspathSnapshot(classpathChanges: ClasspathChanges.ClasspathSnapshotEnabled, lookupStorage: LookupStorage) {
```

Comment:
Logic in shrinkAndSave grew quite a bit, does it make sense to move it out of this calss?

Reply:
Yup - I've moved it to `ClasspathSnapshotShrinker`.